pkg/vsphere/diag/diag.go

```diff
@@ -0,0 +1,149 @@
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// StatusCodeFatalThreshold defines a threshold after which all codes can be treated as fatal.
+const StatusCodeFatalThreshold = 64
+
+const (
+	// VCStatusOK vSphere API is available.
+	VCStatusOK = 0
+	// VCStatusInvalidURL Provided vSphere API URL is wrong.
+	VCStatusInvalidURL = 64
+	// VCStatusErrorQuery Error happened trying to query vSphere API
+	VCStatusErrorQuery = 65
+	// VCStatusErrorResponse Received response doesn't contain expected data.
+	VCStatusErrorResponse = 66
+	// VCStatusIncorrectResponse Received in case if returned data from server is different from expected.
+	VCStatusIncorrectResponse = 67
+	// VCStatusNotXML Received response is not XML
+	VCStatusNotXML = 68
+	// VCStatusUnknownHost is returned in case if DNS failed to resolve name.
+	VCStatusUnknownHost = 69
+	// VCStatusHostIsNotReachable
+	VCStatusHostIsNotReachable = 70
+)
+
+// UserReadableVCAPITestDescription convert API test code into user readable text
+func UserReadableVCAPITestDescription(code int) string {
+	switch code {
+	case VCStatusOK:
+		return "vSphere API target responds as expected"
+	case VCStatusInvalidURL:
+		return "vSphere API target url is invalid"
+	case VCStatusErrorQuery:
+		return "vSphere API target failed to respond to the query"
+	case VCStatusIncorrectResponse:
+		return "vSphere API target returns unexpected response"
+	case VCStatusErrorResponse:
+		return "vSphere API target returns error"
+	case VCStatusNotXML:
+		return "vSphere API target returns non XML response"
+	case VCStatusUnknownHost:
+		return "vSphere API target can not be resolved from VCH"
+	case VCStatusHostIsNotReachable:
+		return "vSphere API target is out of reach. Wrong routing table?"
+	default:
+		return "vSphere API target test returned unknown code"
+	}
+}
+
+// CheckAPIAvailability accesses vSphere API to ensure it is a correct end point that is up and running.
+func CheckAPIAvailability(targetURL string) int {
+	op := trace.NewOperation(context.Background(), "api test")
+	errorCode := VCStatusErrorQuery
+
+	u, err := url.Parse(targetURL)
+	if err != nil {
+		return VCStatusInvalidURL
+	}
+
+	u.Path = "/sdk/vimService.wsdl"
+	apiURL := u.String()
+
+	op.Debugf("Checking access to: %s", apiURL)
+
+	for attempts := 5; errorCode != VCStatusOK && attempts > 0; attempts-- {
+
+		c := http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+			// Is 20 seconds enough to receive any response from vSphere target server?
+			Timeout: time.Second * 20,
+		}
+		errorCode = queryAPI(op, c.Get, apiURL)
+	}
+	return errorCode
+}
+
+func queryAPI(op trace.Operation, getter func(string) (*http.Response, error), apiURL string) int {
+	resp, err := getter(apiURL)
+	if err != nil {
+		errTxt := err.Error()
+		op.Errorf("Query error: %s", err)
+		if strings.Contains(errTxt, "no such host") {
```

Comment:
Is it possible to actually check for errno's here instead of comparing error strings?